python/cog/server/helpers.py

```diff
@@ -148,8 +180,98 @@ def run(self) -> None:
+                        debug("drain event set")
                         self.drain_event.set()
                         drain_tokens_seen = 0
+
+    async def switch_to_async(self) -> None:
+        """
+        This function is called when the main thread switches to being async.
+        It ensures that the behavior stays the same, but write_hook is only called
+        from the main thread.
+
+        1. Open each stream as a StreamReader.
+        2. Create a task for each stream that will process the results.
+        3. write_hook is called for each complete log line.
+        4. Drain and terminate tokens are handled correctly.
+        5. Once the async tasks are started, shut down the thread.
+
+        We must not call write_hook twice for the same data during the switch.
+        """
+        debug("switch async, drain")
+        # Drain the streams to ensure all buffered data is processed
+        try:
+            self.drain()
+        except RuntimeError:
+            debug("drain failed")
+            raise
+        debug("drain done, shutdown")
+
+        # Shut down the thread
+        # we do this before starting a coroutine that will also read from the same fd
+        # so that shutdown can find the terminate tokens correctly
+        self.shutdown()
+        self.stream_tasks = []
+        self.is_async = True
+        debug("set is async")
+
+        for stream in self._streams:
+            # Open each stream as a StreamReader
+            fd = stream.wrapped.fileno()
+            reader = await async_fdopen(fd)
+
+            # Create a task for each stream to process the results
+            task = asyncio.create_task(self.process_stream(stream, reader))
+            self.stream_tasks.append(task)
+
+        # give the tasks a chance to start
+        await asyncio.sleep(0)
+
+    async def process_stream(
+        self, stream: WrappedStream, reader: asyncio.StreamReader
+    ) -> None:
+        debug("process_stream", stream.name)
+        buffer = io.StringIO()
+        drain_tokens_seen = 0
+        should_exit = False
+
+        async for line in reader:
+
+            if not line:
+                break
+
+            line = line.decode()
+            debug("redirector saw", line)
+
+            if not line.endswith("\n"):
+                buffer.write(line)
+                continue
+
+            full_line = buffer.getvalue() + line.strip()
+
+            # Reset buffer
+            buffer = io.StringIO()
+
+            if full_line.endswith(self.terminate_token):
+                full_line = full_line[: -len(self.terminate_token)]
+                should_exit = True
+
+            if full_line.endswith(self.drain_token):
+                drain_tokens_seen += 1
+                full_line = full_line[: -len(self.drain_token)]
+
+            if full_line:
+                # Call write_hook from the main thread
+                self._write_hook(stream.name, stream.original, full_line + "\n")
```

Comment:
Worth profiling, but precomputed / fixed lengths instead of a string suffix and `len` could be noticeably faster if this is a hot code path.